jax/abstract_arrays.py

```diff
@@ -248,3 +255,167 @@ def _make_concrete_python_scalar(x):
+def eval_polymorphic_shape(shape, values_dict):
+  return tuple(dim.evaluate(values_dict) if type(dim) is Poly else dim
+               for dim in shape)
+
+def _ensure_poly(p):
+  if type(p) is Poly:
+    return p
+
+  return Poly({Mon(): p})
+
+class Poly(dict):
+  """Polynomial with integer coefficients,
+  usable as element in a polymorphic shape.
+
+  type Poly = Map Mon Int -- monomials to coeffs
+  type Mon = Map Str Int
+  """
+
+  def __init__(self, coeffs):
+    # Makes sure Polynomials are always in canonical form to simplify operators:
+    coeffs = {mon: op.index(coeff) for mon, coeff in coeffs.items() if coeff != 0}
+    coeffs = {Mon(): 0} if len(coeffs) == 0 else coeffs
+    super().__init__(coeffs)
+
+  def __add__(self, other):
+    coeffs = self.copy()
+
+    for mon, coeff in _ensure_poly(other).items():
+      coeffs[mon] = coeffs.get(mon, 0) + coeff
+
+    return Poly(coeffs)
+
+  def __sub__(self, other):
+    return self + -other
+
+  def __neg__(self):
+    return Poly({mon: -coeff for mon, coeff in self.items()})
+
+  def __mul__(self, other):
+    coeffs = dict()
+    for (mon1, coeff1), (mon2, coeff2) \
+            in product(self.items(), _ensure_poly(other).items()):
+      mon = mon1 * mon2
+      coeffs[mon] = coeffs.get(mon, 0) + coeff1 * coeff2
+
+    return Poly(coeffs)
+
+  def __rmul__(self, other):
+    return self * other
+
+  def __radd__(self, other):
+    return self + other
+
+  def __rsub__(self, other):
+    return self + -other
+
+  def __floordiv__(self, divisor):
+    q, _ = divmod(self, divisor)  # pytype: disable=wrong-arg-types
+    return q
+
+  def __mod__(self, divisor):
+    _, r = divmod(self, divisor)  # pytype: disable=wrong-arg-types
+    return r
+
+  def __divmod__(self, divisor):
+    if self.is_constant:
+      return divmod(int(self), divisor)
+
+    def divided(count):
+      q, r = divmod(count, divisor)
+      if r != 0:
+        raise ValueError('shapecheck currently only supports strides '
+                         'that exactly divide the strided axis length.')
+      return q
+
+    return Poly(
+      {k: coeff // divisor if k.degree == 0 else divided(coeff)
+       for k, coeff in self.items()}), self.get(Mon(), 0) % divisor
+
+  def __hash__(self):
+    return hash(super())
+
+  def __eq__(self, other):
+    return super().__eq__(_ensure_poly(other))
+
+  def __ne__(self, other):
+    return not self == other
+
+  def __ge__(self, other):
+    other = _ensure_poly(other)
+
+    if other.is_constant and self.is_constant:
+      return int(self) >= int(other)
+
+    if other.is_constant and int(other) <= 1:
+      # Assume polynomials > 0, allowing to use shape rules of binops, conv:
+      return True
+
+    if self.is_constant and int(self) <= 0:
+      return False  # See above.
+
+    if self == other:
+      return True
+
+    raise ValueError('Polynomials comparison "{} >= {}" is inconclusive.'
+                     .format(self, other))
+
+  def __le__(self, other):
```

Comment:
You might be able to use `functools.total_ordering` instead of writing out all these explicitly.

Reply:
@mattjj Tried that: It breaks comparison with integers (needed for many shape rules).